iceoryx_posh/source/popo/waitset/wait_set.cpp

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+#include "iceoryx_posh/internal/popo/waitset/wait_set.hpp"
+
+namespace iox
+{
+namespace popo
+{
+WaitSet::WaitSet(cxx::not_null<ConditionVariableData* const> condVarDataPtr) noexcept
+    : m_conditionVariableDataPtr(condVarDataPtr)
+    , m_conditionVariableWaiter(m_conditionVariableDataPtr)
+    , m_guardCondition(m_conditionVariableDataPtr)
+{
+    m_conditionVector.push_back(&m_guardCondition);
+}
+
+bool WaitSet::attachCondition(Condition& condition) noexcept
+{
+    if (!condition.isConditionVariableAttached())
+    {
+        if (condition.attachConditionVariable(m_conditionVariableDataPtr))
+        {
+            return m_conditionVector.push_back(&condition);
+        }
+    }
+
+    return false;
+}
+
+bool WaitSet::detachCondition(Condition& condition) noexcept
+{
+    for (auto currentCondition : m_conditionVector)
+    {
+        if (currentCondition == &condition)
+        {
+            m_conditionVector.erase(&currentCondition);
+            return true;
+        }
+    }
+    return false;
+}
+
+void WaitSet::clear() noexcept
+{
+    m_conditionVector.clear();
+}
+
+cxx::vector<Condition*, MAX_NUMBER_OF_CONDITIONS>
+WaitSet::waitAndReturnFulfilledConditions(bool enableTimeout, units::Duration timeout) noexcept
+{
+    cxx::vector<Condition*, MAX_NUMBER_OF_CONDITIONS> conditionsWithFulfilledPredicate;
+
+    /// @note Inbetween here and last wait someone could have set the trigger to true, hence reset it
+    m_conditionVariableWaiter.reset();
+
+    // Is one of the conditons true?
+    for (auto currentCondition : m_conditionVector)
+    {
+        if (currentCondition->hasTrigger())
+        {
+            conditionsWithFulfilledPredicate.push_back(currentCondition);
+            currentCondition->resetTrigger();
+        }
+    }
+
+    if (conditionsWithFulfilledPredicate.empty())
+    {
+        if (enableTimeout)
+        {
+            auto retVal = m_conditionVariableWaiter.timedWait(timeout);
+
+            if (retVal == false)
+            {
+                // Return empty list
+                return conditionsWithFulfilledPredicate;
+            }
+        }
+        else
+        {
+            m_conditionVariableWaiter.wait();
+        }
+
+        // Check again if one of the conditions is true after we received the signal
+        for (auto currentCondition : m_conditionVector)
```

Comment:
If you're ok with it, I'd like to keep the loop twice. See comment above regarding our use-case.

Reply:
@mossmaurice please follow always the DRY (dont repeat yourself) principle. The reason is that in the code above is a potential bug and with this duplicated code it is possible that the bug is only fixed once and sticks in the second code repetition.

Reply:
What happens if the same condition is true twice this could result in:
- the same condition contained multiple times in `conditionsWithFulfilledPredicate`
- this could lead to an overflow in `conditionsWithFulfilledPredicate` 

We either set the capacity of `conditionsWithFulfilledPredicate` to ` MAX_NUMBER_OF_CONDITIONS * 2` to avoid the overflow or we check if the condition is already contained inside the vector which would solve both problems.

Additionally, we need a unit test to verify the correct behavior of this special case

Reply:
Ok, sleeping about it again, I see your point :-) Refactored it with a lambda and added the test case.